clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -24,123 +24,380 @@ package common
+		if stop {
+			return
+		}
+		funcF := f.(*func())
+		(*funcF)()
+		w.queue.Done(f)
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) gather(c clientset.Interface, syncTimeout time.Duration) error {
+	glog.Infof("%v: waiting for controlled pods measurement...", waitForControlledPodsRunningName)
+	if w.informer == nil {
+		return fmt.Errorf("metric %s has not been started", waitForControlledPodsRunningName)
+	}
+	startTime := time.Now()
+	desiredCount, maxResourceVersion, err := w.getObjectCountAndMaxVersion(c)
 	if err != nil {
-		return summaries, err
+		return err
+	}
+
+	isUpToDate := false
+	for ; time.Since(startTime) < syncTimeout; time.Sleep(checkControlledPodsInterval) {
+		isUpToDate = w.opResourceVersion >= maxResourceVersion
+		if isUpToDate {
+			break
+		}
+	}
+	if !isUpToDate {
+		return fmt.Errorf("timed out while waiting for controlled pods")
 	}
-	labelSelector, err := util.GetStringOrDefault(config.Params, "labelSelector", "")
+
+	w.handlingGroup.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	numberRunning := 0
+	for _, checker := range w.checkerMap {
+		if status := checker.getStatus(); status {
+			numberRunning++
+		}
+	}
+	if desiredCount != numberRunning {
+		glog.Errorf("%s: incorrect objects number: %d/%d %ss are running with all pods", waitForControlledPodsRunningName, numberRunning, desiredCount, w.kind)
+		return fmt.Errorf("incorrect objects number")
+	}
+
+	glog.Infof("%s: %d/%d %ss are running with all pods", waitForControlledPodsRunningName, numberRunning, desiredCount, w.kind)
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) stop() {
+	close(w.stopCh)
+	w.queue.ShutDown()
+	w.workerGroup.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.isRunning = false
+	for _, checker := range w.checkerMap {
+		checker.terminate()
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) handleObject(c clientset.Interface, oldObj, newObj interface{}) {
+	var oldRuntimeObj runtime.Object
+	var newRuntimeObj runtime.Object
+	var ok bool
+	oldRuntimeObj, ok = oldObj.(runtime.Object)
+	if oldObj != nil && !ok {
+		return
+	}
+	defer func() {
+		if err := w.updateOpResourceVersion(oldRuntimeObj); err != nil {
+			glog.Errorf("%s: updating resource version error: %v", waitForControlledPodsRunningName, err)
+		}
+	}()
+	newRuntimeObj, ok = newObj.(runtime.Object)
+	if newRuntimeObj != nil && !ok {
+		return
+	}
+	defer func() {
+		if err := w.updateOpResourceVersion(newRuntimeObj); err != nil {
+			glog.Errorf("%s: updating resource version error: %v", waitForControlledPodsRunningName, err)
+		}
+	}()
+	if isEqual, err := runtimeobjects.IsEqualRuntimeObjectsSpec(oldRuntimeObj, newRuntimeObj); err != nil || isEqual {
+		return
+	}
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	if !w.isRunning {
+		return
+	}
+
+	if oldRuntimeObj != nil {
+		if err := w.deleteObjectLocked(c, oldRuntimeObj); err != nil {
+			glog.Errorf("%s: delete checker error: %v", waitForControlledPodsRunningName, err)
+		} else {
```

Comment:
} else if newRuntimeObj == nil {

[If possible, we want to avoid nesting]

Reply:
Done.